src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -438,6 +434,80 @@ AvroDeserializer::DeserializeFn AvroDeserializer::createDeserializeFn(const avro
                         target_type->getName());
                 }
             }
+
+            if (target.isVariant())
+            {
+                const auto & variant_type = assert_cast<const DataTypeVariant &>(*target_type);
+                const auto & nested_types = variant_type.getVariants();
+
+                using AvroUnionIndex = size_t;
+                std::map<AvroUnionIndex, ColumnVariant::Discriminator> union_index_to_global_discriminator;
+                std::vector<DeserializeFn> nested_deserializers;
+                nested_deserializers.reserve(root_node->leaves());
+
+                bool union_has_null = false;
+                for (size_t i = 0; i != root_node->leaves(); ++i)
+                {
+                    const auto & avro_node = root_node->leafAt(static_cast<int>(i));
+                    if (avro_node->type() == avro::AVRO_NULL)
+                    {
+                        union_has_null = true;
+                        nested_deserializers.emplace_back();
+                        union_index_to_global_discriminator.insert_or_assign(i, ColumnVariant::NULL_DISCRIMINATOR);
+                        continue;
+                    }
+                    const auto variant = AvroSchemaReader::avroNodeToDataType(avro_node);
+                    nested_deserializers.emplace_back(createDeserializeFn(avro_node, variant));
+                    union_index_to_global_discriminator.insert_or_assign(i, variant_type.tryGetVariantDiscriminator(variant->getName()).value());
+                }
+
+                if (root_node->leaves() != nested_types.size() + (union_has_null ? 1 : 0))
+                    throw Exception(ErrorCodes::INCORRECT_DATA, "The number of (non-null) union types in Avro record ({}) doesn't match the number of types in destination Variant type ({}).", root_node->leaves() - (union_has_null ? 1 : 0), nested_types.size());
+
+                if (!union_has_null)
```

Comment:
I think it's better to remove this part. Optimization is tiny, but this can lead to the problems with code support in the future

Reply:
Is it really tiny? The deserializer code runs for each row at the input. We plan to ingest +10 million messages per second from Kafka using this and would like to avoid as many unnecessary instructions as possible (we don't have `null` in our union schema).

Reply:
@Avogar, what do you think?

Reply:
@Avogar, what do you think?

Reply:
My opinion is that branch prediction will help you in the case when there are no NULL entries in data, but I want to ask more experienced colleagues about this :)

Reply:
I think we can have single deserializer and have condition
```c++
...
if (union_has_null && global_discriminator == ColumnVariant::NULL_DISCRIMINATOR)
{
    ...
}
...
```
With `union_has_null = false` I think it will be optimized by branch prediction and we won't see any performance difference

Reply:
Yes, thanks. @jirislav, could you rewrite the code in this way, please?